keps/sig-cluster-lifecycle/kubeadm/4656-add-kubelet-instance-configuration/README.md

```diff
@@ -0,0 +1,287 @@
+* kubeadm currently adds an annotation with the key `kubeadm.alpha.kubernetes.io/cri-socket` to each Node object. We will deprecate and remove it.
+* Provide an instance configuration file named `/var/lib/kubelet/instace-config.yaml` for each node, in which the `ContainerRuntimeEndpoint` field is defined. During the `kubeadm init/join/upgrade` process, the instance configuration will be read and the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` will be overwritten.
+* The `--container-runtime-endpoint` flag is no longer written to the `/var/lib/kubelet/kubeadm-flags.env` file
+
+### Non-Goals
+
+- Continue maintaining CRI socket paths on Node objects.
+
+## Proposal
+
+kubeadm init: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instance-config.yaml` configuration file based on it; if the CRI socket is not specified, the container runtime endpoint will be automatically detected, uploaded to the global configuration, and `/var/lib/kubelet/instace-config.yaml` will be generated.
+
+kubeadm join: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it, overwriting the kubelet configuration downloaded from the global configuration; if the CRI socket is not specified, the kubelet configuration downloaded from the global configuration is used, and `/var/lib/kubelet/instace-config.yaml` is generated based on it.
+
+kubeadm upgrade:  future versions of `kubeadm upgrade apply/node` will only check `/var/lib/kubelet/instace-config.yaml`.
+
+### Risks and Mitigations
+
+## Design Details
+
+**Alpha(v1.31):** 
+
+kubeadm init:
+
+* No longer need to write the `--container-runtime-endpoint` to `/var/lib/kubelet/kubeadm-flags.env`.
+* No longer need to add the `kubeadm.alpha.kubernetes.io/cri-socket` annotation.
+* If the CRI socket provided in the kubeadm configuration is set, it is used first and the `/var/lib/kubelet/instace-config.yaml` configuration file is generated based on it. If the CRI socket is not set, the container runtime endpoint is automatically detected uploaded to the global configuration, and `/var/lib/kubelet/instace-config.yaml` will be generated.
+
+kubeadm join:
+
+* No longer need to add the `kubeadm.alpha.kubernetes.io/cri-socket` annotation.
+* If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it, overwriting the kubelet configuration downloaded from the global configuration; if the CRI socket is not set, the kubelet configuration downloaded from the global configuration is used, and `/var/lib/kubelet/instace-config.yaml` is generated based on it.
+
+kubeadm reset:
+
+* There is no need to do anything, as `/var/lib/kubelet/instace-config.yaml` will be cleaned up with the deletion of `/var/lib/kubelet.`
```

Comment:
something to note here, the file will be deleted but it must be deleted only after we use the socket from the file because that is how we are going to connect to the CRI API to delete running containers.

is that a problem?

Reply:
This is not a problem for us because we first set `resetData.criSocketPath` and then remove the `/var/lib/kubelet` directory

I will explain it in more detail